crates/ibc/src/context/nft_transfer.rs

```diff
@@ -0,0 +1,318 @@
+//! IBC Non-Fungible token transfer context
+
+use std::cell::RefCell;
+use std::rc::Rc;
+
+use namada_core::ibc::apps::nft_transfer::context::{
+    NftTransferExecutionContext, NftTransferValidationContext,
+};
+use namada_core::ibc::apps::nft_transfer::types::error::NftTransferError;
+use namada_core::ibc::apps::nft_transfer::types::{
+    ClassData, ClassUri, Memo, PrefixedClassId, TokenData, TokenId, TokenUri,
+    PORT_ID_STR,
+};
+use namada_core::ibc::core::handler::types::error::ContextError;
+use namada_core::ibc::core::host::types::identifiers::{ChannelId, PortId};
+use namada_core::types::address::Address;
+use namada_core::types::ibc::{NftClass, NftMetadata, IBC_ESCROW_ADDRESS};
+use namada_core::types::token::Amount;
+
+use super::common::IbcCommonContext;
+use crate::storage;
+
+/// NFT transfer context to handle tokens
+#[derive(Debug)]
+pub struct NftTransferContext<C>
+where
+    C: IbcCommonContext,
+{
+    inner: Rc<RefCell<C>>,
+}
+
+impl<C> NftTransferContext<C>
+where
+    C: IbcCommonContext,
+{
+    /// Make new NFT transfer context
+    pub fn new(inner: Rc<RefCell<C>>) -> Self {
+        Self { inner }
+    }
+}
+
+impl<C> NftTransferValidationContext for NftTransferContext<C>
+where
+    C: IbcCommonContext,
+{
+    type AccountId = Address;
+    type Nft = NftMetadata;
+    type NftClass = NftClass;
+
+    fn get_port(&self) -> Result<PortId, NftTransferError> {
+        Ok(PORT_ID_STR.parse().expect("the ID should be parsable"))
+    }
+
+    fn can_send_nft(&self) -> Result<(), NftTransferError> {
+        Ok(())
+    }
+
+    fn can_receive_nft(&self) -> Result<(), NftTransferError> {
+        Ok(())
+    }
+
+    /// Validates that the NFT can be created or updated successfully.
+    fn create_or_update_class_validate(
+        &self,
+        class_id: &PrefixedClassId,
+        _class_uri: Option<&ClassUri>,
+        _class_data: Option<&ClassData>,
+    ) -> Result<(), NftTransferError> {
+        match self.get_nft_class(class_id) {
+            Err(NftTransferError::NftClassNotFound) => Ok(()),
+            Err(e) => Err(e),
+            Ok(class) if class.class_id != *class_id => {
```

Comment:
Given that we query for that specific `class_id` can it happen that we retrieve an `NftClass` with a different class id?

Reply:
Thanks, it's redundant because `get_nft_class` always returns the same `class_id`.